### ISSUE TITLE
static float f = 0.f / 0.f; cannot be compiled

### ISSUE THREAD

**Issue by tydeman** [llvm/llvm-project#161806]
This program will not compile.

int main(void){
  if(1){
    static float f = 0.f / 0.f;/* cannot compile this constant l-value expression yet */
  }
  return 0;
}

I believe that C99 introduced this feature along with IEEE-754 floating-point.
It is one way to create a NAN.

**Comment by fuhsnn** [comment]
The case is explicitly listed in C standard Annex F:
> F.8.5 Constant expressions
> 1. An arithmetic constant expression of floating type, other than one in an initializer for an object
that has static or thread storage duration or that is declared with storage-class specifier constexpr,
is evaluated (as if) during execution ...
> 2. EXAMPLE
```C
#include <fenv.h>
#pragma STDC FENV_ACCESS ON
void f(void)
{
  constexpr double v = 0.0/0.0; // does not raise an exception
  float w[] = { 0.0/0.0 }; // raises an exception
  static float x = 0.0/0.0; // does not raise an exception
  float y = 0.0/0.0; // raises an exception
  double z = 0.0/0.0; // raises an exception
}
```
> 3. For the static and constexpr initializations, the division is done at translation time, raising no (execution-
time) floating-point exceptions.

The example code can be compiled with `gcc -std=c23 -pedantic` but `clang-21` fails with the same error.

**Comment by AaronBallman** [comment]
https://godbolt.org/z/Y56W1ao17

NaN is in the range of representable values (we conform to as much of Annex F as we can), so this should be accepted IMO. CC @zahiraam who has done some work in this area. Curiously enough, at file scope it works fine, seems to only be an issue at local scope.

**Comment by zahiraam** [comment]
This is my understanding of it.
When `static` is used, the initializer must be a constant expression evaluated at compile time. In C,  `0.0f/0.0f` is not a valid constant expression because division by `0` is `UB` in constant expression.
Without the `static`, the initializer is evaluated at run time; `0.0f/0.0f` produces `NaN` and this compiles according the `IEEE754`.
